File: OJ practices/LeetCode题目集/105. Construct Binary Tree from Preorder and Inorder Traversal(medium).java
/*
 * @Author: SourDumplings
 * @Date: 2020-09-04 09:59:06
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
 */

class Solution
{
    public TreeNode buildTree(int[] preorder, int[] inorder)
    {
        if (preorder.length == 0)
        {
            return null;
        }
        int rootVal = preorder[0];
        int leftEnd = inorder.length;
        for (int i = 0; i < inorder.length; i++)
        {
            if (inorder[i] == rootVal)
            {
                leftEnd = i;
                break;
            }
        }
        TreeNode root = new TreeNode(rootVal);
        root.left = build(preorder, 1, leftEnd + 1, inorder, 0, leftEnd);
        root.right = build(preorder, leftEnd + 1, preorder.length, inorder, leftEnd + 1,
            inorder.length);
        return root;
    }

    private TreeNode build(int[] preOrder, int preB, int preE, int[] inOrder, int inB, int inE)
    {
        if (preE <= preB)
        {
            return null;
        }
        int rootVal = preOrder[preB];
        TreeNode root = new TreeNode(rootVal);
        int leftEnd = inE;
        for (int i = inB; i < inE; i++)
        {
            if (inOrder[i] == rootVal)
            {
                leftEnd = i;
                break;
            }
        }
        root.left = build(preOrder, preB + 1, leftEnd - inB + preB + 1, inOrder, inB, leftEnd);
        root.right = build(preOrder, leftEnd - inB + preB + 1, preE, inOrder, leftEnd + 1, inE);
        return root;
    }
}

File: OJ practices/LeetCode题目集/105. Construct Binary Tree from Preorder and Inorder Traversal(medium).cpp
/**
 * @file 105. Construct Binary Tree from Preorder and Inorder Traversal(medium).cpp
 * @author SourDumplings (sourdumplings@qq.com)
 * @brief
 * https://leetcode.cn/problems/construct-binary-tree-from-preorder-and-inorder-traversal/?envType=study-plan-v2&envId=top-interview-150
 * @version 1.0.0
 * @date 2024-12-09
 *
 * @copyright Copyright (c) 2024 SourDumplings
 *
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
	TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder)
	{
		size_t n = preorder.size();
		return buildTree(preorder, 0, n, inorder, 0, n);
	}

private:
	TreeNode *buildTree(
		const vector<int> &pre,
		size_t preB,
		size_t preE,
		const vector<int> &in,
		size_t inB,
		size_t inE
	)
	{
		if (preB == preE) { return nullptr; }
		int rootVal	   = pre[preB];
		TreeNode *root = new TreeNode(rootVal);
		for (size_t i = inB; i < inE; i++)
		{
			if (in[i] == rootVal)
			{
				root->left	= buildTree(pre, preB + 1, preB + 1 + i - inB, in, inB, i);
				root->right = buildTree(pre, preB + 1 + i - inB, preE, in, i + 1, inE);
				break;
			}
		}
		return root;
	}
};